### PR TITLE
fix: fix incorrect zkiterator start position

### PR DESCRIPTION
**Pull request by Pangssu**

### CODE REVIEW

**Review by sm-stack — APPROVED** [review]
LGTM

File: trie/iterator.go [hunk 1]
```diff
@@ -967,51 +964,125 @@ func (it *merkleTreeIterator) seek(path []byte) {
...
+				it.stack = it.stack[:len(it.stack)-1]
+			case right: // right visited. go left
+				if leftNode := it.resolveNode(it.parentOfLastNode().children[left]); leftNode != nil {
+					it.path[lastPathIndex] = left
+					it.stack[len(it.stack)-1] = leftNode
+					it.visitBeforeLeafNode()
+					return true
+				}
+				if it.err != nil {
+					return false
```
Comment (seolaoh) [inline trie/iterator.go:1058]: If an error occurred, `prevLeaf` returns false. Don't we need to add `if it.err != nil { return }` in the caller function `seek`?

File: trie/iterator_test.go [hunk 2]
```diff
@@ -18,20 +18,25 @@ package trie
...
+	"golang.org/x/exp/slices"
+
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
 	"github.com/ethereum/go-ethereum/core/rawdb"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/ethdb/memorydb"
 	"github.com/ethereum/go-ethereum/trie/trienode"
 	"github.com/ethereum/go-ethereum/trie/zk"
```
Comment (seolaoh) [inline trie/iterator_test.go:39]: Maybe this is the right order?
```go
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/common/math"
	"github.com/ethereum/go-ethereum/core/rawdb"
	"github.com/ethereum/go-ethereum/core/types"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/ethdb/memorydb"
	"github.com/ethereum/go-ethereum/trie/trienode"
	"github.com/ethereum/go-ethereum/trie/zk"
	"golang.org/x/exp/slices"

	zkt "github.com/kroma-network/zktrie/types"
```

File: trie/iterator_test.go [hunk 3]
```diff
@@ -722,4 +727,210 @@ func TestMerkleTreeIterator(t *testing.T) {
...
+	// 60605793585936038718222944469367330975045253165943558504994436015126663057264
+	// 85820665183479862833099943033748518022104258034302536812924589654960629631092
+	// 85903213226554376758871722034242209534889372863809613891381864212832081120192
+	// 90325294576998166929651976356705470786861187287245759150875554730994004964376
+	// 100426110323909178507910617048938140138509083918485403644527157931210594668280
+	// 101121973535399381299596552991112472380608726703043018240306005545403524541156
+	// 107728238818287576212658976290069527420728087670830176810004760597624147961456
+	// 108965031955345082684882074272074576895157534795165059799486221336737027828244
+	// 109990652738555876101121666461374586270194093281384129526042736168378582559940
+	// 110211460199408811616376306299384222697381109719033443765762911326375974320200
```
Comment (seolaoh) [inline trie/iterator_test.go:833]: Please add a comment what it is.

File: trie/iterator.go [hunk 4]
```diff
@@ -946,17 +946,14 @@ func newMerkleTreeIterator(
...
 
-func (it *merkleTreeIterator) seek(path []byte) {
-	if len(path) == 0 {
-		return
-	}
-	path = zk.NewTreePathFromHashBig(common.BytesToHash(path))
-
+func (it *merkleTreeIterator) seek(path zk.TreePath) {
 	for _, p := range path {
 		if parent, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
```
Comment (seolaoh) [inline trie/iterator.go:958]: I think you should check if `len(it.stack) > 0` before this line.

File: trie/iterator.go [hunk 5]
```diff
@@ -967,51 +964,125 @@ func (it *merkleTreeIterator) seek(path []byte) {
...
+				if it.err != nil {
+					return
+				}
+			}
+		} else {
+			break
+		}
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply trie/iterator.go:1042]: Agree

File: trie/iterator.go [hunk 6]
```diff
@@ -967,51 +964,125 @@ func (it *merkleTreeIterator) seek(path []byte) {
...
+func (it *merkleTreeIterator) prevLeaf() bool {
+	if _, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorLeafNode); ok {
```
Comment (seolaoh) [inline trie/iterator.go:1044]: I think you should check if `len(it.stack) > 0` before this line.

Comment (seolaoh) [reply trie/iterator.go:1044]: Also I think all other places should have this check. like `Hash()`, `NodeBlob()`, `Leaf()`, `parentOfLastNode()`, `lastNodeAsLeaf()`

File: trie/iterator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jyc228) [reply trie/iterator.go:1042]: If the stack length is 0, it indicates that the iteration has ended.
In most cases, the stack length should never reach 0, nor should it be allowed to.
Returning in such a state would make it impossible to verify if it was an abnormal termination.
Given this, I recommend keeping the current implementation as is.

If you agree, I will add more detailed comments to the stack structure.

File: trie/iterator.go (hunk 4 above)
Comment (jyc228) [reply trie/iterator.go:958]: Feedback on stack length 0: https://github.com/kroma-network/go-ethereum/pull/118#discussion_r1855684834

File: trie/iterator.go (hunk 6 above)
Comment (jyc228) [reply trie/iterator.go:1044]: Feedback on stack length 0: https://github.com/kroma-network/go-ethereum/pull/118#discussion_r1855684834

File: trie/iterator_test.go (hunk 2 above)
Comment (jyc228) [reply trie/iterator_test.go]: updated

File: trie/iterator.go (hunk 1 above)
Comment (jyc228) [reply trie/iterator.go:1064]: updated

File: trie/iterator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (seolaoh) [reply trie/iterator.go:1048]: @jyc228 ok, please add more comment

File: trie/iterator.go [hunk 7]
```diff
@@ -967,51 +964,131 @@ func (it *merkleTreeIterator) seek(path []byte) {
...
+		case 0: // leaf path == path
+		case 1: // leaf path < path
+		loop2:
+			for it.nextLeaf() {
+				if it.err != nil {
+					return
+				}
+				switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(it.lastNodeAsLeaf().key))) {
+				case -1: // path < leaf path
+					it.prevLeaf()
```
Comment (seolaoh) [inline trie/iterator.go:1002]: I think we need to check `it.err != nil` after call `it.prevLeaf()`
```suggestion
				switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(it.lastNodeAsLeaf().key))) {
				case -1: // path < leaf path
					it.prevLeaf()
					if it.err != nil {
						return
					}
```

File: trie/iterator.go [hunk 8]
```diff
@@ -967,51 +965,131 @@ func (it *merkleTreeIterator) seek(path []byte) {
...
-		} else {
-			break
 		}
+		break
 	}
 
-	if len(it.path) == 0 {
-		return // root node is not parent node
+	if _, ok := it.stack[len(it.stack)-1].(*merkleTreeIteratorParentNode); ok {
+		it.nextLeaf()
```
Comment (seolaoh) [inline trie/iterator.go:973]:
```suggestion
		it.nextLeaf()
		if it.err != nil {
			return
		}
```

File: trie/iterator.go [hunk 9]
```diff
@@ -967,51 +965,131 @@ func (it *merkleTreeIterator) seek(path []byte) {
...
+				if it.err != nil {
+					return
+				}
+				switch bytes.Compare(path, zk.NewTreePathFromHashBig(common.BytesToHash(it.lastNodeAsLeaf().key))) {
+				case -1: // path < leaf path
+					continue
+				case 0: // path == leaf path
+					break loop1
+				case 1:
+					it.nextLeaf()
```
Comment (seolaoh) [inline trie/iterator.go:990]:
```suggestion
					it.nextLeaf()
					if it.err != nil {
						return
					}
```

**Review by seolaoh — APPROVED** [review]

**Review by kangsorang — APPROVED** [review]

**Pull request merged**